File: python/Trees/validate_binary_search_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
        Brute Force Approach:
        - For each node, check that all nodes in its left subtree are strictly less than the node's value,
          and all nodes in its right subtree are strictly greater than the node's value.
        - This is done by traversing the left and right subtrees for every node.
        - Time Complexity: O(N^2) in the worst case (for each node, traverse its subtrees).
        - Space Complexity: O(H) for recursion stack, where H is the height of the tree.
        """
        def check_left(node, val):
            if not node:
                return True
            if node.val >= val:
                return False
            return check_left(node.left, val) and check_left(node.right, val)

        def check_right(node, val):
            if not node:
                return True
            if node.val <= val:
                return False
            return check_right(node.left, val) and check_right(node.right, val)

        def is_bst(node):
            if not node:
                return True
            # Check all nodes in left subtree are < node.val
            if not check_left(node.left, node.val):
                return False
            # Check all nodes in right subtree are > node.val
            if not check_right(node.right, node.val):
                return False
            # Recursively check for left and right subtrees
            return is_bst(node.left) and is_bst(node.right)

        return is_bst(root)
        
class Solution2:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        """
        Efficient Approach (Bounds Checking):
        - For each node, ensure its value is strictly between the allowed lower (left) and upper (right) bounds.
        - Recursively check the left subtree with an updated upper bound (current node's value),
          and the right subtree with an updated lower bound (current node's value).
        - If any node violates the BST property, return False.
        - Time Complexity: O(N), where N is the number of nodes (each node is visited once).
        - Space Complexity: O(H), where H is the height of the tree (recursion stack).
        """
        def valid(node, left, right):
            if not node:
                return True
            # Node's value must be strictly between left and right bounds
            if not (left < node.val < right):
                return False
            # Recursively check left and right subtrees with updated bounds
            return (valid(node.left, left, node.val) and
                    valid(node.right, node.val, right))

        return valid(root, float('-inf'), float('inf'))

# -------------------
# Helper Functions for Testing
# -------------------
def build_tree_from_list(lst):
    if not lst:
        return None
    nodes = [TreeNode(val) if val is not None else None for val in lst]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root

# -------------------
# Test Cases
# -------------------
def test_is_valid_bst():
    s = Solution2()
    # Test case 1: Example
    root = build_tree_from_list([2,1,3])
    assert s.isValidBST(root) == True
    # Test case 2: Not a BST
    root = build_tree_from_list([5,1,4,None,None,3,6])
    assert s.isValidBST(root) == False
    # Test case 3: Single node
    root = build_tree_from_list([1])
    assert s.isValidBST(root) == True
    # Test case 4: Empty tree
    root = build_tree_from_list([])
    assert s.isValidBST(root) == True
    # Test case 5: Duplicates not allowed
    root = build_tree_from_list([2,2,2])
    assert s.isValidBST(root) == False
    # Test case 6: Valid BST with negative values
    root = build_tree_from_list([0,-1])
    assert s.isValidBST(root) == True
    print("All test cases passed!")

if __name__ == "__main__":
    test_is_valid_bst()

File: C++/Trees/validate_binary_search_tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    /**
     * Efficient Approach:
     * - Validate that each node's value is strictly between given bounds.
     * - Left subtree must have all nodes < current node.
     * - Right subtree must have all nodes > current node.
     * 
     * Time Complexity: O(N), each node visited once.
     * Space Complexity: O(H), where H is the height of the tree (recursion stack).
     */
    bool isValidBST(TreeNode* root) {
        return valid(root, LONG_MIN, LONG_MAX);
    }

private:
    bool valid(TreeNode* node, long left, long right) {
        if (!node) return true;

        // Node's value must lie between left and right bounds
        if (!(left < node->val && node->val < right)) return false;

        // Recursively check left and right subtrees with updated bounds
        return valid(node->left, left, node->val) &&
               valid(node->right, node->val, right);
    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers.

using namespace std;
    ✅ Allows writing std::vector, std::cout etc., without std:: prefix.

struct TreeNode {
    ✅ Defines the structure for a binary tree node.

int val;
    ✅ Integer value of this node.

TreeNode *left, *right;
    ✅ Pointers to left and right children.

TreeNode() : val(0), left(nullptr), right(nullptr) {}
    ✅ Default constructor: value 0, children nullptr.

TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    ✅ Constructor that sets value and nullptr children.

TreeNode(int x, TreeNode *left, TreeNode *right)
    ✅ Constructor that sets value and given left/right children.

class Solution {
    ✅ Defines the Solution class.

public:
    ✅ Public functions accessible outside the class.

bool isValidBST(TreeNode* root)
    ✅ Function that starts the recursion with the entire tree and extreme bounds.

return valid(root, LONG_MIN, LONG_MAX);
    ✅ Calls helper with root and bounds (-∞, +∞) (LONG_MIN and LONG_MAX).

private:
    ✅ Private helper function only used internally.

bool valid(TreeNode* node, long left, long right)
    ✅ Recursive function that checks if the subtree rooted at `node` is a valid BST
       within the bounds (`left`, `right`).

if (!node) return true;
    ✅ Base case: null nodes are valid BSTs.

if (!(left < node->val && node->val < right)) return false;
    ✅ If node violates the BST property, return false.

return valid(node->left, left, node->val) &&
       valid(node->right, node->val, right);
    ✅ Recur on left subtree with updated upper bound (current value).
       Recur on right subtree with updated lower bound (current value).
       Both must be valid.

*/
